src/main/java/me/jellysquid/mods/sodium/mixin/features/block/MixinBlockModelRenderer.java

```diff
@@ -30,9 +34,14 @@ public class MixinBlockModelRenderer {
      * @reason Use optimized vertex writer intrinsics, avoid allocations
      * @author JellySquid
      */
-    @Overwrite
-    public void render(MatrixStack.Entry entry, VertexConsumer vertexConsumer, BlockState blockState, BakedModel bakedModel, float red, float green, float blue, int light, int overlay) {
-        var writer = VertexBufferWriter.of(vertexConsumer);
+    @Inject(method = "render(Lnet/minecraft/client/util/math/MatrixStack$Entry;Lnet/minecraft/client/render/VertexConsumer;Lnet/minecraft/block/BlockState;Lnet/minecraft/client/render/model/BakedModel;FFFII)V", at = @At("HEAD"), cancellable = true)
+    private void renderFast(MatrixStack.Entry entry, VertexConsumer vertexConsumer, BlockState blockState, BakedModel bakedModel, float red, float green, float blue, int light, int overlay, CallbackInfo ci) {
+        var writer = VertexConsumerUtils.convertOrLog(vertexConsumer);
+        if(writer == null) {
+            return;
+        }
+
+        ci.cancel();
```

Comment:
Isn't this just going to always cancel? Shouldn't this only cancel on when the writer is null?

Reply:
No. Immediately above this line the function returns without cancelling when the writer is null.

Reply:
I thought `CallbackInfo#cancel` is effectively a premature method return.

Reply:
The injection into the target method works like this:

```java
public void originalMethod(int x) {
    CallbackInfo ci = new CallbackInfo();
    this.handler$inject$rrr000(x, ci);
    if(ci.isCancelled())
        return;
}
```

so it does not matter whether you cancel at the beginning or the end inside the injected method.

Reply:
In this case we cancel when we have a non-null writer as Sodium handles the rendering itself, but skip cancelling for a null writer as we want the vanilla logic below the injected handler to run.